doc/decisions/0002-use-heroku-run-detached-for-static-site-build.md

```diff
@@ -0,0 +1,43 @@
+# 2. Use heroku run:detached for static site build
+
+Date: 2018-12-05
+
+## Status
+
+Accepted
+
+## Context
+
+At the moment, we have a set of scheduled tasks on Heroku that run every 10 minutes, hour, or day. These range from building the static site (every 10 minutes), to reporting stalled builds (every hour), to refreshing the database from production data (in dev/staging, every night). The main job here, building the static site, has started to fail because it occasionally runs longer than 10 minutes. The Heroku scheduler will only allow a scheduled instance to live for as long as the interval on the task. Heroku is therefore killing the instance before it can finish, leading to the incomplete builds and frequent reports of 'stalled' (in this instance, killed) builds.
+
+We need to schedule static site builds in a way that removes this strict time constraint so that they will always finish building.
+
+## Decision
+
+We will continue to use the Heroku scheduler on a ten-minute interval, but will instead use the Heroku CLI to run the management command (./manage.py build_static_site) in a separate, detached worker dyno. This worker dyno has a lifespan of 24 hours and is not limited to the interval of the scheduled job.
+
+The command we will use is `heroku run:detached -a <app_name> ./manage.py build_static_site`.
+
+Pros
+----
+* Very easy to implement.
+* Will allow static site builds that take up to 24 hours (which is greater than any reasonable build length).
+
+Cons
+----
+* We remain unable to responsively schedule static site builds at the moment they are needed and rely on the scheduler to kick off a build.
+* We have to expose a Heroku API key and install the Heroku CLI buildpack on our main CMS app.
```

Comment:
I think it's also worth acknowledging that this would also be the first time that we include some code in our app which ties us to a particular infrastructure provider - everything else is generic and transferrable. That feels like a significant downside to me, although may be a pragmatic short-term fix.

Reply:
I don't feel like we're really including any _code_ that ties us to a provider. We're adding one env variable and a line to `app.json`, which is already a heroku-specific file?

Reply:
Ah, I misunderstood. I'd assumed that we'd be calling the `heroku run:detached` command from within our code, but that's actually just setup within the scheduler app itself.

In which case, the only downside is having the Heroku CLI in the buildpack, which feels like it's ok.